### PR TITLE
feat: support application function

### PR DESCRIPTION
**Pull request by artsiomkorzun**

### CODE REVIEW

**Comment by Oleksii-Klimov** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-504.nightly-test.deltixhub.io
>E2E tests status: success

**Comment by artsiomkorzun** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-504.nightly-test.deltixhub.io
>E2E tests status: success

**Comment by avavilau** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-504.nightly-test.deltixhub.io
>E2E tests status: success

File: src/main/java/com/epam/aidial/core/controller/ApplicationController.java [hunk 1]
```diff
@@ -68,12 +79,85 @@ public Future<?> getApplicationService() {
...
+    public Future<?> startApplication() {
+        context.getRequest()
+                .body()
+                .compose(body -> {
+                    String url = ProxyUtil.convertToObject(body, ResourceLink.class).url();
+                    ResourceDescription resource = decodeUrl(url);
+                    checkAccess(resource);
+                    return vertx.executeBlocking(() -> applicationService.startApplication(context, resource), false);
+                })
+                .onSuccess(publication -> context.respond(HttpStatus.OK, publication))
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/controller/ApplicationController.java]: rename publication -> application

File: src/main/java/com/epam/aidial/core/controller/ApplicationController.java [hunk 2]
```diff
@@ -68,12 +79,85 @@ public Future<?> getApplicationService() {
...
+    public Future<?> stopApplication() {
+        context.getRequest()
+                .body()
+                .compose(body -> {
+                    String url = ProxyUtil.convertToObject(body, ResourceLink.class).url();
+                    ResourceDescription resource = decodeUrl(url);
+                    checkAccess(resource);
+                    return vertx.executeBlocking(() -> applicationService.stopApplication(resource), false);
+                })
+                .onSuccess(publication -> context.respond(HttpStatus.OK, publication))
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/controller/ApplicationController.java]: the same

File: src/main/java/com/epam/aidial/core/controller/ApplicationController.java [hunk 3]
```diff
@@ -68,12 +79,85 @@ public Future<?> getApplicationService() {
...
+    public Future<?> getApplicationLogs() {
+        context.getRequest()
+                .body()
+                .compose(body -> {
+                    String url = ProxyUtil.convertToObject(body, ResourceLink.class).url();
+                    ResourceDescription resource = decodeUrl(url);
+                    checkAccess(resource);
+                    return vertx.executeBlocking(() -> applicationService.getApplicationLogs(resource), false);
+                })
+                .onSuccess(publication -> context.respond(HttpStatus.OK, publication))
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/controller/ApplicationController.java]: rename publication -> logs

File: src/main/java/com/epam/aidial/core/controller/DeploymentPostController.java [hunk 4]
```diff
@@ -153,6 +157,9 @@ private void handleRequestError(String deploymentId, Throwable error) {
         } else if (error instanceof ResourceNotFoundException) {
             log.error("Deployment not found {}", deploymentId, error);
             respond(HttpStatus.NOT_FOUND, error.getMessage());
+        } else if (error instanceof HttpException e) {
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/controller/DeploymentPostController.java:167]: Might be better to replace with the block as you did in the application controller?
See `private void respondError(Throwable error)`

File: src/main/java/com/epam/aidial/core/security/AccessService.java [hunk 5]
```diff
@@ -226,8 +228,16 @@ private Map<ResourceDescription, Set<ResourceAccessType>> getReviewAccess(
...
 
         return resources.stream()
-                .filter(resource -> PublicationService.isReviewResource(resource)
-                        && PublicationService.hasReviewAccess(context, resource))
+                .filter(resource -> PublicationService.hasReviewAccess(context, resource))
+                .collect(Collectors.toUnmodifiableMap(
+                        Function.identity(), resource -> ResourceAccessType.READ_ONLY));
+    }
+
+    private Map<ResourceDescription, Set<ResourceAccessType>> getDeploymentAccess(
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/security/AccessService.java:236]: It's not clear the method is named as `getDeploymentAccess` but it checks if user has permission to a resource in the review status
What's the purpose of the method?

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java [hunk 6]
```diff
@@ -18,32 +20,97 @@
...
+
+        this.vertx = vertx;
+        this.httpClient = httpClient;
         this.encryptionService = encryptionService;
         this.resourceService = resourceService;
-        this.controllerUrl = settings.getString("controllerUrl", null);
+        this.lockService = lockService;
+        this.idGenerator = idGenerator;
+        this.pendingApplications = redis.getScoredSortedSet(pendingApplicationsKey, StringCodec.INSTANCE);
+        this.controllerEndpoint = settings.getString("controllerEndpoint", "https://ai-dial-app-controller.deltixuat.com");
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/service/ApplicationService.java]: Please don't keep any reference to internal resources

Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/service/ApplicationService.java]: Please document application settings

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java [hunk 7]
```diff
@@ -18,32 +20,97 @@
...
+        this.controllerTimeout = settings.getLong("controllerTimeout", 300000L);
+        this.checkDelay = settings.getLong("checkDelay", 300000L);
+        this.checkSize = settings.getInteger("checkSize", 64);
         this.includeCustomApps = settings.getBoolean("includeCustomApps", false);
+
+        long checkPeriod = settings.getLong("checkPeriod", 300000L);
+        vertx.setPeriodic(checkPeriod, checkPeriod, ignore -> vertx.executeBlocking(this::checkApplications));
+    }
+
+    public static boolean hasDeploymentAccess(ProxyContext context, ResourceDescription resource) {
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/service/ApplicationService.java:89]: It's almost the same as hasReviewAccess. The difference is the check: url should contain `deployments`. Why?

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java [hunk 8]
```diff
@@ -155,121 +220,266 @@ public List<Application> getApplications(ResourceDescription resource, Consumer<
...
 
         return Pair.of(meta, application);
     }
 
     public void deleteApplication(ResourceDescription resource, EtagHeader etag) {
-        if (resource.isFolder() || resource.getType() != ResourceType.APPLICATION) {
-            throw new IllegalArgumentException("Invalid application url: " + resource.getUrl());
-        }
+        verifyApplication(resource);
+        AtomicReference<Application> reference = new AtomicReference<>();
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/service/ApplicationService.java]: We can replace AtomicReference with MutableObject

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java [hunk 9]
```diff
@@ -279,4 +489,420 @@ private void prepareApplication(ResourceDescription resource, EtagHeader etag,
...
+
+            if (function == null) {
+                throw new IllegalStateException("Application has no function");
+            }
+
+            if (function.getStatus() != Application.Function.Status.STARTING) {
+                throw new IllegalStateException("Application is not starting");
+            }
+
+            if (!isPublicOrReview(resource)) {
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/service/ApplicationService.java:520]: Why do we not copy source files to target if an app is public?

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java [hunk 10]
```diff
@@ -279,4 +489,420 @@ private void prepareApplication(ResourceDescription resource, EtagHeader etag,
...
+
+            pendingApplications.remove(resource.getUrl());
+            return null;
+        } catch (Throwable error) {
+            log.warn("Failed to launch application: {}", resource.getUrl(), error);
+            throw error;
+        }
+    }
+
+    private void createApplicationImage(ProxyContext context, Application.Function function) {
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/service/ApplicationService.java]: Can we consider move the methods to call k8s stuff to a separate class?

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java [hunk 11]
```diff
@@ -279,4 +489,420 @@ private void prepareApplication(ResourceDescription resource, EtagHeader etag,
...
+            log.warn("Failed to check pending applications:", e);
+        }
+
+        return null;
+    }
+
+    // region Launching Application
+
+    private Void launchApplication(ProxyContext context, ResourceDescription resource) {
+        try (LockService.Lock lock = lockService.tryLock(deploymentLockKey(resource))) {
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/service/ApplicationService.java:502]: The operation may take a while. There is a timeout after that the lock will be released.

What should we do then?

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java [hunk 12]
```diff
@@ -279,4 +489,420 @@ private void prepareApplication(ResourceDescription resource, EtagHeader etag,
...
+
+    private void deleteApplicationDeployment(Application.Function function) {
+        callController(HttpMethod.DELETE, "/v1/deployment/" + function.getId(),
+                request -> null,
+                (response, body) -> convertServerSentEvent(body, EmptyResponse.class));
+    }
+
+    // endregion
+
+    private void copyFolder(String sourceFolderUrl, String targetFolderUrl) {
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/service/ApplicationService.java]: Let's move the method to ResourceService

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java [hunk 13]
```diff
@@ -279,4 +489,420 @@ private void prepareApplication(ResourceDescription resource, EtagHeader etag,
...
+                if (!resourceService.copyResource(sourceFile, targetFile)) {
+                    throw new IllegalStateException("Can't copy function source file: " + sourceFileUrl);
+                }
+            }
+
+            token = folder.getNextToken();
+        } while (token != null);
+    }
+
+    private void deleteFolder(String folderUrl) {
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/service/ApplicationService.java]: Let's move the method to ResourceService

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java [hunk 14]
```diff
@@ -279,4 +489,420 @@ private void prepareApplication(ResourceDescription resource, EtagHeader etag,
...
+
+        String name = encryptionService.encrypt(location);
+        return ResourceDescription.fromDecoded(ResourceType.FILE, name, location, null).getUrl();
+    }
+
+    public static boolean isActive(Application application) {
+        return application != null && application.getFunction() != null && isActive(application.getFunction().getStatus());
+    }
+
+    private static boolean isActive(Application.Function.Status status) {
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/service/ApplicationService.java]: Let's move the method to the class Status

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java [hunk 15]
```diff
@@ -279,4 +489,420 @@ private void prepareApplication(ResourceDescription resource, EtagHeader etag,
...
+            case CREATED, FAILED, STOPPED -> false;
+            case STARTING, STARTED, STOPPING -> true;
+        };
+    }
+
+    private static boolean isPending(Application application) {
+        return application != null && application.getFunction() != null && isPending(application.getFunction().getStatus());
+    }
+
+    private static boolean isPending(Application.Function.Status status) {
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/service/ApplicationService.java]: the same

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java [hunk 16]
```diff
@@ -279,4 +489,420 @@ private void prepareApplication(ResourceDescription resource, EtagHeader etag,
...
+        } catch (Throwable e) {
+            throw new IllegalArgumentException(message);
+        }
+    }
+
+    private static String buildMapping(String endpoint, String path) {
+        return (endpoint == null || path == null) ? null : (endpoint + path);
+    }
+
+    private static <T> T convertServerSentEvent(String body, Class<T> clazz) {
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/service/ApplicationService.java]: Let's move to utility class and cover with tests

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java [hunk 17]
```diff
@@ -155,121 +220,266 @@ public List<Application> getApplications(ResourceDescription resource, Consumer<
...
             if (function.getSourceFolder() == null) {
-                throw new IllegalArgumentException("Application function sources must be provided");
+                throw new IllegalArgumentException("Application function source folder must be provided");
             }
 
             try {
                 ResourceDescription folder = ResourceDescription.fromAnyUrl(function.getSourceFolder(), encryptionService);
 
-                if (!folder.isFolder() || folder.getType() != ResourceType.FILE) {
+                if (!folder.isFolder() || folder.getType() != ResourceType.FILE || !folder.getBucketName().equals(resource.getBucketName())) {
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/service/ApplicationService.java:465]: Empty error message

File: src/main/java/com/epam/aidial/core/service/PublicationUtil.java [hunk 18]
```diff
@@ -50,43 +50,6 @@ public String replaceConversationLinks(String conversationBody, ResourceDescript
         return conversation.toString();
     }
 
-    public String replaceApplicationLinks(String body, ResourceDescription targetResource, boolean preserveReference, Map<String, String> attachmentsMapping) {
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/service/PublicationUtil.java:53]: I suppose the new logic is not the same as old.
you lost `preserve` for example

File: src/main/java/com/epam/aidial/core/util/ProxyUtil.java [hunk 19]
```diff
@@ -272,12 +266,12 @@ public static <T> T convertToObject(String payload, TypeReference<T> type) {
     }
 
     @Nullable
-    public static <T> T convertToObject(String payload, Class<T> clazz, boolean snakeCase) {
+    public static <T> T convertToObject(String payload, Class<T> clazz) {
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/util/ProxyUtil.java:269]: Did you replace the arg with Json annotations?

File: src/main/java/com/epam/aidial/core/controller/ApplicationController.java (hunk 1 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/controller/ApplicationController.java]: Done

File: src/main/java/com/epam/aidial/core/controller/ApplicationController.java (hunk 2 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/controller/ApplicationController.java]: done

File: src/main/java/com/epam/aidial/core/controller/ApplicationController.java (hunk 3 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/controller/ApplicationController.java]: done

File: src/main/java/com/epam/aidial/core/controller/DeploymentPostController.java (hunk 4 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/controller/DeploymentPostController.java:167]: respond finalizes request, I do not want to add one more method to finilize request and send this error using Context::respond.

File: src/main/java/com/epam/aidial/core/security/AccessService.java (hunk 5 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/security/AccessService.java:236]: It uses ApplicationService.hasDeploymentAccess to check if an user has access to special deployment bucket.

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java (hunk 6 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/service/ApplicationService.java]: Test-only, will remove before merging.

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java (hunk 7 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/service/ApplicationService.java:89]: To have different path on storage. 
* Review - /Users/User/publications/id/resource_type
* Deployment - /Users/User/deployments/id/files

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java (hunk 8 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/service/ApplicationService.java]: done

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java (hunk 9 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/service/ApplicationService.java:520]: Sources are copied to an immutable and hidden folder for a review/public application. So we do not need to create a temporary folder when starting such applications.

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java (hunk 11 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/service/ApplicationService.java:502]: This timeout is currently less than lock timeout. So I decided not to implement lock watchdog mechanism in this mr.

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java (hunk 17 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/service/ApplicationService.java:465]: the following catch block throws a generic error with message, so this message won't be used.

File: src/main/java/com/epam/aidial/core/service/PublicationUtil.java (hunk 18 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/service/PublicationUtil.java:53]: By default it is preserved. Publication code generates a new reference.

File: src/main/java/com/epam/aidial/core/util/ProxyUtil.java (hunk 19 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/util/ProxyUtil.java:269]: Now snake case is defined with the annotation.

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java (hunk 15 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/service/ApplicationService.java]: done

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java (hunk 16 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/service/ApplicationService.java]: Already covered. Moved.

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java (hunk 14 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/service/ApplicationService.java]: done

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java (hunk 13 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/service/ApplicationService.java]: done

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java (hunk 12 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/service/ApplicationService.java]: done

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java (hunk 10 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/service/ApplicationService.java]: moved

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java (hunk 11 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/service/ApplicationService.java:502]: Please add a comment about your decision was made.

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java (hunk 7 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/service/ApplicationService.java:89]: Should we introduce a new resource type then?

File: src/main/java/com/epam/aidial/core/service/ApplicationOperator.java [hunk 20]
```diff
@@ -0,0 +1,222 @@
...
+import java.util.Map;
+import java.util.StringTokenizer;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+class ApplicationOperator {
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/service/ApplicationOperator.java]: It's worth adding java doc. It's not clear what the service does

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java (hunk 7 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/service/ApplicationService.java:89]: The source files should be stored as files to be available via file API.

File: src/main/java/com/epam/aidial/core/service/ApplicationOperator.java (hunk 20 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/service/ApplicationOperator.java]: added

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java (hunk 11 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/service/ApplicationService.java:502]: added

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java (hunk 6 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/service/ApplicationService.java]: added

**Review by astsiapanay — DISMISSED** [review]

**Comment by Gimir** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-504.nightly-test.deltixhub.io
>E2E tests status: success

File: src/main/java/com/epam/aidial/core/service/ApplicationService.java (hunk 6 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/service/ApplicationService.java]: Done

File: src/main/java/com/epam/aidial/core/AiDial.java [hunk 21]
```diff
@@ -157,11 +156,9 @@ void stop() {
             close(storage);
             close(redis);
             log.info("Proxy stopped");
-            LogConfigurator.unconfigure();
         } catch (Throwable e) {
             log.warn("Proxy failed to stop:", e);
-            LogConfigurator.unconfigure();
-            System.exit(-1);
+            throw e;
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/AiDial.java:161]: Why did you remove these lines?

**Review by astsiapanay — APPROVED** [review]

File: src/main/java/com/epam/aidial/core/AiDial.java (hunk 21 above)
Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/AiDial.java:161]: I moved it to the main method. start/stop should not change global settings and should not have global-like behavior.

Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/AiDial.java:161]: AiDial instances are used in tests. So  `LogConfigurator.unconfigure()` disabled all logging for the following tests
